staging/src/k8s.io/apiextensions-apiserver/pkg/controller/apiapproval/apiapproval_controller.go

```diff
@@ -0,0 +1,262 @@
+func calculateCondition(crd *apiextensions.CustomResourceDefinition) *apiextensions.CustomResourceDefinitionCondition {
+	if !apihelpers.IsGroupProtected(crd.Spec.Group) {
+		return nil
+	}
+
+	switch approvalState := apihelpers.GetAPIApprovalState(crd.Annotations); approvalState {
+	case apihelpers.APIApprovalInvalid:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved: protected groups must have approval approval with either URL or unapproved, see https://github.com/kubernetes/enhancements/pull/1111", crd.Name),
+		}
+	case apihelpers.APIApprovalMissing:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved: protected groups must have approval, see https://github.com/kubernetes/enhancements/pull/1111", crd.Name),
+		}
+	case apihelpers.APIApproved:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v was approved in %v", crd.Name, crd.Annotations[v1beta1.KubeAPIApprovalAnnotation]),
+		}
+	case apihelpers.APIUnapproved:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved: %q", crd.Name, crd.Annotations[v1beta1.KubeAPIApprovalAnnotation]),
+		}
+	default:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionUnknown,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v unrecognized approval state %v; coding error", crd.Name, approvalState),
+		}
+	}
+}
+
+func (c *KubeAPIApprovedConditionController) sync(key string) error {
+	inCustomResourceDefinition, err := c.crdLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// avoid repeated calculation for the same generation
+	c.lastSeenGenerationLock.Lock()
+	lastSeen, seenBefore := c.lastSeenGeneration[inCustomResourceDefinition.Name]
+	c.lastSeenGenerationLock.Unlock()
+	if seenBefore && inCustomResourceDefinition.Generation <= lastSeen {
+		return nil
+	}
+
+	// check old condition
+	cond := calculateCondition(inCustomResourceDefinition)
+	if cond == nil {
+		return nil
+	}
+	old := apiextensions.FindCRDCondition(inCustomResourceDefinition, apiextensions.KubeAPIApproved)
+
+	// don't attempt a write if all the condition details are the same
+	if old != nil && old.Status == cond.Status && old.Reason == cond.Reason && old.Message == cond.Message {
+		return nil
```

Comment:
should we update the generation in that case?

Reply:
added comment